src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy

```diff
@@ -0,0 +1,115 @@
+import org.dormitory.autobotsoccub.engine.model.GameData
+import spock.lang.Specification
+
+import static org.dormitory.autobotsoccub.engine.model.PlayPosition.GK
+
+class GameEngineShould extends Specification {
+
+    def GK_A = 0
+    def FW_A = 1
+    def GK_B = 2
+    def FW_B = 3
+
+    Game game
+    GameEngine sut = new GameEngine()
+
+    def setup() {
+        game = Game.builder()
+                .goalKeeperA(GK_A)
+                .forwardA(FW_A)
+                .goalKeeperB(GK_B)
+                .forwardB(FW_B)
+                .build()
+    }
+
+    def "correctly started game"() {
+        when:
+        def newGame = sut.startGame(FW_A, game)
+        def existGame = sut.startGame(FW_A, game)
+
+        then:
+        newGame
+        !existGame
+    }
+
+    def "correctly executed game commands"() {
+        given:
+        sut.startGame(FW_A, game)
+        sut.metaClass.getCurrentGame << { Integer userId ->
+            GameData activeGame = activeGames.get userId
+            GameData.builder()
+                    .game(activeGame.getGame())
+                    .scoreTable(activeGame.getScoreTable())
+                    .build()
+        }
+
+        //correctly count score && count auto score && changed position in the team
+        when:
+        sut.score FW_A
+        sut.autoScore FW_A
+        sut.changePositionsInTeam FW_A
+        def forwardAStats = sut.getCurrentGame(FW_A).scoreTable.statsByUserId.get(FW_A)
+
+        then:
+        forwardAStats.scored == 1
+        forwardAStats.autoScored == 1
+        forwardAStats.position == GK
+
+        //correctly catch exception
+        when:
+        sut.score 999
+
+        then:
+        GameDoesNotExistException e = thrown()
+        e.cause == null
+
+        //correctly reverted last event & stop game
+        when:
+        sut.score FW_B
+        sut.revert FW_B
+        sut.autoScore FW_B
+        sut.revert FW_B
+        sut.revert FW_B //check the situation without event
+        def gameData = sut.stopGame FW_B
+
+        then:
+        gameData.game != null
+        gameData.scoreTable != null
+        def forwardBStats = gameData.scoreTable.statsByUserId.get(FW_B)
+        forwardBStats.scored == 0
+        forwardBStats.autoScored == 0
+    }
+
+    def "tested game"() {
+        given:
+        sut.startGame(FW_A, game)
+
+        when:
+        sut.score FW_A
+        sut.score FW_A
+        sut.score FW_A
+        sut.autoScore FW_A
+        sut.score FW_B
+        sut.score FW_B
+        sut.revert FW_B
+        sut.changePositionsInTeam GK_A
+        sut.score FW_B
+        sut.score GK_B
+
+        def userStats = sut.stopGame(FW_A).scoreTable.statsByUserId
+        def forwardA = userStats.get(FW_A)
+        def goalkeeperA = userStats.get(GK_A)
+        def forwardB = userStats.get(FW_B)
+        def goalkeeperB = userStats.get(GK_B)
+
+        then:
+        forwardA.scored == 3    && forwardA.autoScored == 1     && forwardA.missed == 2
```

Comment:
use dataTable to specify your expectations

Reply:
I deleted this method, 'cause the engine needs to be more developed for valuable game simulation.